File: java/src/main/java/com/workshop/architecture/fitness/membership/infrastructure/MembershipEntity.java
package com.workshop.architecture.fitness.membership.infrastructure;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

@Entity
@Table(name = "memberships")
public class MembershipEntity {

    @Id
    private UUID id;

    @Column(nullable = false)
    private String customerId;

    @Column(nullable = false)
    private String planId;

    @Column(nullable = false)
    private int planPrice;

    @Column(nullable = false)
    private int planDuration;

    @Column(nullable = false)
    private String status;

    @Column
    private String reason;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @Column
    private LocalDate pauseStartDate;

    @Column
    private LocalDate pauseEndDate;

    @Column
    private String pauseReason;

    @Column
    private Instant cancelledAt;

    @Column
    private String cancellationReason;

    protected MembershipEntity() {
    }

    public MembershipEntity(
            UUID id,
            String customerId,
            String planId,
            int planPrice,
            int planDuration,
            String status,
            String reason,
            LocalDate startDate,
            LocalDate endDate
    ) {
        this.id = id;
        this.customerId = customerId;
        this.planId = planId;
        this.planPrice = planPrice;
        this.planDuration = planDuration;
        this.status = status;
        this.reason = reason;
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public UUID getId() {
        return id;
    }

    public String getCustomerId() {
        return customerId;
    }

    public String getPlanId() {
        return planId;
    }

    public int getPlanPrice() {
        return planPrice;
    }

    public int getPlanDuration() {
        return planDuration;
    }

    public String getStatus() {
        return status;
    }

    public String getReason() {
        return reason;
    }

    public LocalDate getStartDate() {
        return startDate;
    }

    public LocalDate getEndDate() {
        return endDate;
    }

    public LocalDate getPauseStartDate() {
        return pauseStartDate;
    }

    public LocalDate getPauseEndDate() {
        return pauseEndDate;
    }

    public String getPauseReason() {
        return pauseReason;
    }


}

File: python/src/workshop_api/fitness/membership_entity.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Column, Date, DateTime, Integer, String

from workshop_api.fitness.database import Base


class MembershipOrmModel(Base):
    __tablename__ = "memberships"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    customer_id = Column(String(36), nullable=False)
    plan_id = Column(String(36), nullable=False)
    plan_price = Column(Integer, nullable=False)
    plan_duration = Column(Integer, nullable=False)
    status = Column(String(50), nullable=False)
    reason = Column(String(100), nullable=True)
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=False)
    pause_start_date = Column(Date, nullable=True)
    pause_end_date = Column(Date, nullable=True)
    pause_reason = Column(String(100), nullable=True)
    cancelled_at = Column(DateTime(timezone=True), nullable=True)
    cancellation_reason = Column(String(100), nullable=True)


class MembershipBillingReferenceOrmModel(Base):
    __tablename__ = "membership_billing_references"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    membership_id = Column(String(36), nullable=False)
    external_invoice_id = Column(String(100), nullable=False, unique=True)
    external_invoice_reference = Column(String(100), nullable=False, unique=True)
    due_date = Column(Date, nullable=False)
    status = Column(String(50), nullable=False)
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
